source/core/module_eventHelpers.js

```diff
@@ -27,22 +27,18 @@
     // Static Private Variables (all vireo instances)
 
     var assignEventHelpers = function (Module, publicAPI) {
-        // Disable new-cap for the cwrap functions so the names can be the same in C and JS
-        /* eslint 'new-cap': ['error', {'capIsNewExceptions': [
-            'OccurEvent'
-        ]}], */
-
         Module.eventHelpers = {};
         publicAPI.eventHelpers = {};
 
-        var OccurEvent = Module.cwrap('OccurEvent', 'void', ['number', 'number', 'number']);
-
         var registerForControlEvent = function () {
             throw new Error('No event registration callback was supplied');
         };
         var unRegisterForControlEvent = function () {
             throw new Error('No event un-registration callback was supplied');
         };
+        var writeEventData = function () {
```

Comment:
This callback is used to write the data. In the PR for the dynamic variable allocation, this was a method passed occurEvent, but I believe all of the callbacks would be the same. Any event could just use the Vireo poker to generically write JS objects into the allocated memory (VIREO_POKER.pokeValueRef(vireo, valueRef, eventData). Therefore, I'm making this callback something that can be set once for Vireo like the register and unregister methods.

Reply:
I think it should be something that is specified per occurEvent call. There can be other code paths and event types that may occur an event besides value properties and I don't think we necessarily should force them down a single function registration.

One example is that we have discussed implementing WebSockets using Events which would receive strings over the network. While strings that go through the peeker and poker are always assumed to be JS strings, we can actually get a UTF-8 array buffer directly from a websocket and have a more optimized copy. Forcing one write function means we have to twist it to support all possible use cases for events occurring.

I would actually be in favor of switching from a function being passed to a visitor being passed. That would keep the api interface consistent between other api functions we added like `reflectOnValueRef (typeVisitor, valueRef, data)`

If we were to do that we can change the occurEvent signature to be:
`occurEvent = function (eventOracleIndex, controlId, eventType, typeVisitor, eventDataTypeValueRef, eventData)`

And in the occurEvent function do:
``` 
occurEvent = function {
var allocatedDataValueRef = ...
Module.eggShell.reflectOnValueRef(typeVisitor, valueRef, data);
Module._occurEvent(...)
Module.eggShell.deallocateData(...)
}
```

I say we either switch to passing in a writeFunction every time or passing in a typeVisitor every time, but I'm not for one writefunction / typeVisistor being used for all calls.

Reply:
I've updated it to take a visitor and use reflectOnValueRef